src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,254 @@
+#pragma once
+#include <vector>
+#include "storage/projected_columns.h"
+#include "storage/projected_row.h"
+#include "storage/write_ahead_log/log_io.h"
+
+namespace terrier::storage {
+// TODO(Zhaozhe): Should fetch block_size. Use magic number first.
+// The block size to output to disk every time
+#define CHECKPOINT_BLOCK_SIZE 4096
+
+/**
+ * The header of a page in the checkpoint file.
+ */
+// TODO(Zhaozhe, Mengyang): More fields can be added to header
+class CheckpointFilePage {
+ public:
+  /**
+   * Initialize a given page
+   * @param page to be initialized.
+   */
+  static void Initialize(CheckpointFilePage *page) {
+    // TODO(mengyang): support non-trivial initialization
+    page->checksum_ = 0;
+    page->table_oid_ = catalog::table_oid_t(0);
+  }
+
+  /**
+   * Get the checksum of the page.
+   * @return checksum of the page.
+   */
+  uint32_t GetChecksum() { return checksum_; }
+
+  /**
+   * Set the checksum of the page.
+   * @param checksum
+   */
+  void SetCheckSum(uint32_t checksum) { checksum_ = checksum; }
+
+  /**
+   * Get the oid of the table whose rows are stored in this page.
+   * @return oid of the table.
+   */
+  catalog::table_oid_t GetTableOid() { return table_oid_; }
+
+  /**
+   * Set the oid of the table whose rows are stored in this page.
+   * @param oid of the table.
+   */
+  void SetTableOid(catalog::table_oid_t oid) { table_oid_ = oid; }
+
+  /**
+   * Get the pointer to the first row in this page.
+   * @return pointer to the first row.
+   */
+  byte *GetPayload() { return varlen_contents_; }
+
+ private:
+  uint32_t checksum_;
+  catalog::table_oid_t table_oid_;
+  byte varlen_contents_[0];
+};
+
+/**
+ * A buffered tuple writer collects tuples into blocks, and output a block once a time.
+ * The block size is fetched from database setting, and should be the page size.
+ * layout:
+ * ----------------------------------------------------------------------------------
+ * | checksum | table_oid | tuple1 | varlen_entries(if exist) | tuple2 | tuple3 | ... |
+ * ----------------------------------------------------------------------------------
+ */
+class BufferedTupleWriter {
+  // TODO(Zhaozhe): checksum
+ public:
+  BufferedTupleWriter() = default;
+
+  /**
+   * Instantiate a new BufferedTupleWriter, open a new checkpoint file to write into.
+   * @param log_file_path path to the checkpoint file.
+   */
+  explicit BufferedTupleWriter(const char *log_file_path)
+      : out_(PosixIoWrappers::Open(log_file_path, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR)),
+        block_size_(CHECKPOINT_BLOCK_SIZE),
+        buffer_(new byte[block_size_]()) {
```

Comment:
You have code replicated here, why can't just use the open function you defined below?